airflow/io/__init__.py

```diff
@@ -0,0 +1,297 @@
+
+        Returns:
+            OutputStream: An object that matches the OutputStream protocol.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+            FileExistsError: If the file at self.location already exists and `overwrite=False`.
+        """
+
+
+class FileIO(ABC):
+    """A base class for FileIO implementations."""
+
+    @abstractmethod
+    def new_input(self, location: str, conn_id: str | None) -> InputFile:
+        """Get an InputFile instance to read bytes from the file at the given location.
+
+        Args:
+            location (str): A URI or a path to a local file.
+            conn_id (str): The connection ID to use when getting the file.
+        """
+
+    @abstractmethod
+    def new_output(self, location: str, conn_id: str | None) -> OutputFile:
+        """Get an OutputFile instance to write bytes to the file at the given location.
+
+        Args:
+            location (str): A URI or a path to a local file.
+            conn_id (str): The connection ID to use when getting the file.
+        """
+
+    @abstractmethod
+    def delete(self, location: str | InputFile | OutputFile, conn_id: str | None) -> None:
+        """Delete the file at the given path.
+
+        Args:
+            location (Union[str, InputFile, OutputFile]): A URI or a path to a local file--if an InputFile
+                instance or an OutputFile instance is provided, the location attribute for that instance is
+                used as the URI to delete.
+            conn_id (str): The connection ID to use when deleting the file.
+
+        Raises:
+            PermissionError: If the file at location cannot be accessed due to a permission error.
+            FileNotFoundError: When the file at the provided location does not exist.
+        """
+
+
+FSSPEC_FILE_IO = "airflow.io.fsspec.FsspecFileIO"
+
+# Mappings from the Java FileIO impl to a Python one. The list is ordered by preference.
+# If an implementation isn't installed, it will fall back to the next one.
+SCHEMA_TO_FILE_IO: dict[str, list[str]] = {
+    "s3": [FSSPEC_FILE_IO],
+    "s3a": [FSSPEC_FILE_IO],
+    "s3n": [FSSPEC_FILE_IO],
+    "gs": [FSSPEC_FILE_IO],
+    "file": [FSSPEC_FILE_IO],
+    # "hdfs": [ARROW_FILE_IO],
+    "abfs": [FSSPEC_FILE_IO],
+    "abfss": [FSSPEC_FILE_IO],
+}
+
+
+def _import_file_io(io_impl: str) -> FileIO | None:
+    try:
+        path_parts = io_impl.split(".")
+        if len(path_parts) < 2:
+            raise ValueError(f"py-io-impl should be full path (module.CustomFileIO), got: {io_impl}")
+        module_name, class_name = ".".join(path_parts[:-1]), path_parts[-1]
+        module = importlib.import_module(module_name)
+        class_ = getattr(module, class_name)
+        return class_()
+    except ModuleNotFoundError:
+        log.warning("Could not initialize FileIO: %s", io_impl)
+        return None
+
+
+def _infer_file_io_from_scheme(path: str) -> FileIO | None:
```

Comment:
This private method seems to be un-used or I am too stupid finding where it is used. :-D

Reply:
you are probably correct. Port stuff.